format.go

```diff
@@ -342,23 +351,46 @@ func (a *unknownAny) ProtoMessage() {
 
 var _ proto.Message = (*unknownAny)(nil)
 
-// RequestParserAndFormatterFor returns a request parser and formatter for the
+// FormatterOptions is a set of flags that are passed to a JSON or text formatter.
+type FormatterOptions struct {
+	EmitJSONDefaultFields bool // FormatJSON only
+	IncludeTextSeparator  bool // FormatText only
+	AllowUnknownFields    bool // FormatJSON only
+}
+
+// RequestParserAndFormatterWithOptions returns a request parser and formatter for the
 // given format. The given descriptor source may be used for parsing message
-// data (if needed by the format). The flags emitJSONDefaultFields and
-// includeTextSeparator are options for JSON and protobuf text formats,
-// respectively. Requests will be parsed from the given in.
-func RequestParserAndFormatterFor(format Format, descSource DescriptorSource, emitJSONDefaultFields, includeTextSeparator bool, in io.Reader) (RequestParser, Formatter, error) {
+// data (if needed by the format).
+// It accepts a set of options. The field EmitJSONDefaultFields and IncludeTextSeparator
+// are options for JSON and protobuf text formats, respectively. The AllowUnknownFields field
+// is a JSON-only format flag.
+// Requests will be parsed from the given in.
+func RequestParserAndFormatterWithOptions(format Format, descSource DescriptorSource, opts FormatterOptions, in io.Reader) (RequestParser, Formatter, error) {
 	switch format {
 	case FormatJSON:
 		resolver := AnyResolverFromDescriptorSource(descSource)
-		return NewJSONRequestParser(in, resolver), NewJSONFormatter(emitJSONDefaultFields, anyResolverWithFallback{AnyResolver: resolver}), nil
+		unmarshaler := jsonpb.Unmarshaler{AnyResolver: resolver, AllowUnknownFields: opts.AllowUnknownFields}
+		return NewJSONRequestParserWithUnmarshaler(in, unmarshaler), NewJSONFormatter(opts.EmitJSONDefaultFields, anyResolverWithFallback{AnyResolver: resolver}), nil
 	case FormatText:
-		return NewTextRequestParser(in), NewTextFormatter(includeTextSeparator), nil
+		return NewTextRequestParser(in), NewTextFormatter(opts.IncludeTextSeparator), nil
 	default:
 		return nil, nil, fmt.Errorf("unknown format: %s", format)
 	}
 }
 
+// RequestParserAndFormatterFor returns a request parser and formatter for the
+// given format. The given descriptor source may be used for parsing message
+// data (if needed by the format). The flags emitJSONDefaultFields and
+// includeTextSeparator are options for JSON and protobuf text formats,
+// respectively. Requests will be parsed from the given in.
+func RequestParserAndFormatterFor(format Format, descSource DescriptorSource, emitJSONDefaultFields, includeTextSeparator bool, in io.Reader) (RequestParser, Formatter, error) {
+	opts := FormatterOptions{
```

Comment:
After moving `opts` to end of arg list for subsequent call, this could be inlined.